### PR TITLE
[react-codemirror] Add types for the change object

### PR DESCRIPTION
**Pull request by rudi-c**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://codemirror.net/doc/manual.html
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/react-codemirror/index.d.ts [hunk 1]
```diff
@@ -1,15 +1,36 @@
-// Type definitions for React Codemirror v0.2.6
+// Type definitions for React Codemirror v0.2.7
```
Comment (velveret) [inline types/react-codemirror/index.d.ts]: there is no v0.2.7.

File: types/react-codemirror/index.d.ts [hunk 2]
```diff
@@ -1,15 +1,36 @@
...
 // Definitions by: Vicky Lai <https://github.com/velveret>
+//                 Rudi Chen <https://github.com/rudi-c>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.3
 
 /// <reference types="react"/>
 /// <reference types="codemirror"/>
 
 declare namespace ReactCodeMirror {
+    interface Pos {
```
Comment (velveret) [inline types/react-codemirror/index.d.ts]: `Pos` and `Change` should be the `CodeMirror` types, not added here.

**Review by velveret — CHANGES_REQUESTED** [review]

**Comment by rudi-c** [comment]
True, true, newbie mistakes! Thanks for the suggestions @velveret, please take a look?

File: types/react-codemirror/index.d.ts [hunk 3]
```diff
@@ -1,6 +1,7 @@
 // Type definitions for React Codemirror v0.2.6
```
Comment (velveret) [inline types/react-codemirror/index.d.ts]: your type change isn't in 0.2.6, it's in 1.0.0. 1.0.0 introduces other changes as well - you should update the typings to match all of 1.0.0's changes if you're going to bump the version.
See https://github.com/JedWatson/react-codemirror/blob/master/HISTORY.md#v100--2017-05-14

**Review by velveret — CHANGES_REQUESTED** [review]

**Comment by rudi-c** [comment]
@velveret That's a good idea, done.

File: types/react-codemirror/index.d.ts [hunk 4]
```diff
@@ -1,22 +1,41 @@
...
+        /** Called when the cursor is moved */
+        onCursorActivity?: (cm: CodeMirror.Editor) => any;
+        /** Called when the editor is focused or loses focus */
+        onFocusChange?: (focused: boolean) => any;
+        /** Called when the editor is scrolled */
+        onScroll?: (scrollInfo: CodeMirror.ScrollInfo) => any;
+        /** Options passed to the CodeMirror instance */
+        options?: CodeMirror.EditorConfiguration;
+        /** (DEPRECATED), use `name` */
+        path?: string;
```
Comment (velveret) [inline types/react-codemirror/index.d.ts]: There's a preserveScrollPosition prop as well

File: types/react-codemirror/index.d.ts [hunk 5]
```diff
@@ -1,22 +1,41 @@
...
-        className?: string; // CSS className for the outer element
-        codeMirrorInstance?: CodeMirror.Editor; // the CodeMirror instance
+        /** Automatically focuses the editor when it is mounted (default false) */
+        autoFocus?: boolean;
+        /** Automatically persist changes to underlying textarea (default false) */
+        autoSave?: boolean;
+        /** Adds a custom CSS class to the editor */
+        className?: string;
+        /** Provides a specific CodeMirror instance (defaults to `require('codemirror')`) */
+        codeMirrorInstance?: CodeMirror.Editor;
```
Comment (velveret) [inline types/react-codemirror/index.d.ts]: this has been changed to a function type - see https://github.com/JedWatson/react-codemirror/pull/54

File: types/react-codemirror/index.d.ts [hunk 6]
```diff
@@ -1,22 +1,41 @@
...
+        /** Provides the default (not changed tracked) value to the editor */
+        defaultValue?: string;
+        /** Set the name of the editor input field */
+        name?: string;
+        /** Called when a change is made */
+        onChange?: (newValue: string, change: CodeMirror.EditorChange) => any;
+        /** Called when the cursor is moved */
+        onCursorActivity?: (cm: CodeMirror.Editor) => any;
```
Comment (velveret) [inline types/react-codemirror/index.d.ts]: prefer `codemirror` over `cm` to match react-codemirror docs and be more descriptive

File: types/react-codemirror/index.d.ts [hunk 7]
```diff
@@ -1,22 +1,41 @@
...
-        onScroll?: (scrollInfo: CodeMirror.ScrollInfo) => any; // called when the editor is scrolled
-        options?: CodeMirror.EditorConfiguration; // options passed to the CodeMirror instance
-        path?: string; // the identifying name for the textarea
-        value?: string; // the editor value
-        className?: string; // CSS className for the outer element
-        codeMirrorInstance?: CodeMirror.Editor; // the CodeMirror instance
+        /** Automatically focuses the editor when it is mounted (default false) */
+        autoFocus?: boolean;
+        /** Automatically persist changes to underlying textarea (default false) */
+        autoSave?: boolean;
```
Comment (velveret) [inline types/react-codemirror/index.d.ts]: this prop is listed in the docs but isn't actually in the library - see https://github.com/JedWatson/react-codemirror/pull/22#issuecomment-320414389. I'm fine with leaving this in for now to match the react-codemirror docs and so we don't have to update again once it's restored.

**Comment by rudi-c** [comment]
@velveret I've made the changes, but I have no idea how to refer to a type that's the default export of a module that's also a merge function-interface type. I just put a function signature that works, but can we merge it like this? Investigating this further would be beyond my needs or interest.

File: types/react-codemirror/react-codemirror-tests.tsx [hunk 8]
```diff
@@ -17,19 +17,25 @@ class CodemirrorTest extends React.Component<React.Props<{}>> {
...
         const onFocusChange = (focused: boolean) => console.log(focused);
         const onScroll = (scrollInfo: CodeMirror.ScrollInfo) => console.log(scrollInfo.top);
-        const codeMirrorInstance = CodeMirror(document.body);
+        const codeMirrorInstance = CodeMirror;
 
         return <div>
             <Codemirror className="test-codemirror"
+                        autoFocus={true}
+                        autoSave={true}
                         codeMirrorInstance={codeMirrorInstance}
```
Comment (velveret) [inline types/react-codemirror/react-codemirror-tests.tsx]: you can remove the `codeMirrorInstance` variable and put `CodeMirror` here

**Review by velveret — APPROVED** [review]
Thanks for making all the changes for 1.0.

**Pull request merged**